### PR TITLE
Entry point largest comp

### LINKED ISSUE #86: Medusa issue: spontaneous splitting
image

# Description

The current method of choosing a starting point is "start with a random tip, and backtrack until reaching a transaction with enough cumulative weight.” We introduced a change in #75 : instead of choosing the first transaction which crosses the threshold, we choose the last one which doesn’t. The intention was to get rid of the frequent “sub-tangle is too big” errors that kept occurring during tip selection.

This change caused the Tangle to split into a bunch of split snakes, as the picture shows. While we’re not sure exactly how it came to be, it is clear why it’s stable: we choose a random tip to backtrack from (at the end of one of these snakes), and we don’t backtrack deep enough to get a starting point which allows merging. At some point this forking transaction is so deep that it’s impossible to start the walk there anyway ("sub-tangle is too big" error).

We might try to “fix” this situation by merging branches, but we would ideally like to find a tip selection method that heals itself in such a situation. As an interim solution, we reverted #75 .

# Potential Solutions

1. *Choose 2 independent entry points*
This is tempting, but probably a bad idea, since it will be a complicated coding effort once we re-introduce the ledger state post z-net.

1. *Selection the single “best” tip.* This will prevent splitting by deterministically starting from the same tip. We suggested two potential criteria:
    1. Always start from the most distant tip (from genesis).
    1. Select the largest connected component: 
        1. Take the graph of transactions from the last X minutes
        1. Find connected components and count their size. 
        1.  Choose a random tip from the biggest component.

1. *Optimized or heuristic tip selection when sub-tangle is too big*. The idea is to find a way to never throw “sub-tangle is too big” errors. We suggested several methods to achieve this:

    1. Unweighted steps: during the random walk take mostly alpha=0 steps (which are cheap), and only compute CW for a small fraction of the steps, while hoping to stay on the main tangle.
    1. Keep old weights semi-updated (by storing the last computed CW for each transaction). This a heuristic approach that assumes weight differences don’t change too much for transactions in the deep Tangle.
    1. Propagate CW to approvees when a new transaction becomes solid, rather than computing CW from scratch upon tip selection. This can be combined with caching (only propagate backwards once a “bin” has filled, which is a heuristic to optimize the propagation), and a cap on how far in the past you propagate (beyond which weights are no longer updated, or very infrequently.

### PR DESCRIPTION
**Pull request by alongalky**
Fixes #86, consolidates #88 #89 #90 into a single class and chooses the starting tip from the largest connected component among recent transactions.

### CODE REVIEW

**Comment by alon-e** [comment]
fixes #95 as well

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsStartingTipSelector.java [hunk 1]
```diff
@@ -1,27 +1,41 @@
...
     private final int maxTransactions;
+    private final int maxInitialTips;
     private final Random random;
+    private final TipsViewModel tipsViewModel;
 
-    public ConnectedComponentsCalculator(Tangle tangle, int maxTransactions) {
-        this.maxTransactions = maxTransactions;
+    public ConnectedComponentsStartingTipSelector(Tangle tangle, int maxTransactions, TipsViewModel tipsViewModel) {
         this.tangle = tangle;
+        this.maxTransactions = maxTransactions;
```
Comment (alon-e) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsStartingTipSelector.java:24]: this sounds like a small number - currently 400, especially given that `maxInitialTips` will be 100, so the CC might be very small.

would be good to consult Clara on this, but my personal bias is sqrt(N)

**Review by alon-e — APPROVED** [review]

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsStartingTipSelector.java (hunk 1 above)
Comment (alongalky) [reply src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsStartingTipSelector.java:24]: Are you suggesting to change
```
this.maxInitialTips = maxTransactions / 4;
```
to
```
this.maxInitialTips = sqrt(maxTransactions);
```
?

Comment (alon-e) [reply src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsStartingTipSelector.java:24]: Yes, well converted to int :)

**Pull request merged**